File: foundational_design_patterns/10_hitl/src/hitl_basic.py
"""
Basic Human-in-the-Loop (HITL) Implementation

This module demonstrates a simple HITL workflow for content generation
with a human approval checkpoint before publishing.
"""

import os
import sys
from pathlib import Path
from typing import Optional
from openai import OpenAI
from dotenv import load_dotenv


# Color codes for terminal output
class Colors:
    HEADER = '\033[95m'
    BLUE = '\033[94m'
    CYAN = '\033[96m'
    GREEN = '\033[92m'
    YELLOW = '\033[93m'
    RED = '\033[91m'
    ENDC = '\033[0m'
    BOLD = '\033[1m'
    UNDERLINE = '\033[4m'


def print_header(text: str):
    """Print a formatted header."""
    print(f"\n{Colors.HEADER}{Colors.BOLD}{'='*80}{Colors.ENDC}")
    print(f"{Colors.HEADER}{Colors.BOLD}{text.center(80)}{Colors.ENDC}")
    print(f"{Colors.HEADER}{Colors.BOLD}{'='*80}{Colors.ENDC}\n")


def print_section(title: str):
    """Print a section title."""
    print(f"\n{Colors.CYAN}{Colors.BOLD}{title}{Colors.ENDC}")
    print(f"{Colors.CYAN}{'-'*len(title)}{Colors.ENDC}")


def print_content(content: str):
    """Print generated content in a box."""
    lines = content.split('\n')
    max_length = max(len(line) for line in lines) + 4

    print(f"\n{Colors.BLUE}+{'-'*max_length}+{Colors.ENDC}")
    for line in lines:
        print(f"{Colors.BLUE}|{Colors.ENDC} {line.ljust(max_length-2)} {Colors.BLUE}|{Colors.ENDC}")
    print(f"{Colors.BLUE}+{'-'*max_length}+{Colors.ENDC}\n")


def print_success(text: str):
    """Print success message."""
    print(f"{Colors.GREEN}{Colors.BOLD}✓ {text}{Colors.ENDC}")


def print_error(text: str):
    """Print error message."""
    print(f"{Colors.RED}{Colors.BOLD}✗ {text}{Colors.ENDC}")


def print_warning(text: str):
    """Print warning message."""
    print(f"{Colors.YELLOW}{Colors.BOLD}⚠ {text}{Colors.ENDC}")


class ContentGenerator:
    """Simple content generator using OpenAI."""

    def __init__(self, api_key: str):
        """Initialize the content generator."""
        self.client = OpenAI(api_key=api_key)

    def generate_content(self, content_type: str, topic: str,
                        additional_instructions: str = "") -> str:
        """Generate content using OpenAI."""
        prompts = {
            "blog": f"Write a short blog post (200-300 words) about: {topic}. {additional_instructions}",
            "email": f"Write a professional email about: {topic}. {additional_instructions}",
            "social": f"Write a social media post (2-3 sentences) about: {topic}. {additional_instructions}",
            "tweet": f"Write a tweet (280 characters max) about: {topic}. {additional_instructions}"
        }

        prompt = prompts.get(content_type, prompts["blog"])

        try:
            response = self.client.chat.completions.create(
                model="gpt-4o-mini",
                messages=[
                    {"role": "system", "content": "You are a professional content writer. Create engaging, clear, and concise content."},
                    {"role": "user", "content": prompt}
                ],
                temperature=0.7,
                max_tokens=500
            )
            return response.choices[0].message.content.strip()
        except Exception as e:
            print_error(f"Failed to generate content: {str(e)}")
            return None


class HITLWorkflow:
    """Basic Human-in-the-Loop workflow for content generation."""

    def __init__(self, generator: ContentGenerator):
        """Initialize the HITL workflow."""
        self.generator = generator
        self.decision_log = []

    def get_human_decision(self, content: str, content_type: str) -> tuple[str, Optional[str]]:
        """
        Present content to human and get approval decision.

        Returns:
            tuple: (decision, feedback) where decision is 'approve', 'reject', or 'modify'
        """
        print_section("🔍 HUMAN REVIEW CHECKPOINT")
        print(f"\n{Colors.BOLD}Content Type:{Colors.ENDC} {content_type.upper()}")
        print(f"{Colors.BOLD}Generated Content:{Colors.ENDC}")
        print_content(content)

        print(f"\n{Colors.YELLOW}{Colors.BOLD}Review Options:{Colors.ENDC}")
        print(f"  {Colors.GREEN}[A]{Colors.ENDC} Approve - Publish the content as-is")
        print(f"  {Colors.RED}[R]{Colors.ENDC} Reject - Discard the content")
        print(f"  {Colors.YELLOW}[M]{Colors.ENDC} Modify - Request changes to the content")

        while True:
            decision = input(f"\n{Colors.BOLD}Your decision [A/R/M]:{Colors.ENDC} ").strip().upper()

            if decision in ['A', 'APPROVE']:
                return 'approve', None
            elif decision in ['R', 'REJECT']:
                reason = input(f"{Colors.BOLD}Reason for rejection (optional):{Colors.ENDC} ").strip()
                return 'reject', reason or "No reason provided"
            elif decision in ['M', 'MODIFY']:
                feedback = input(f"{Colors.BOLD}What changes would you like?{Colors.ENDC} ").strip()
                if feedback:
                    return 'modify', feedback
                else:
                    print_warning("Please provide feedback for modifications.")
            else:
                print_error("Invalid option. Please choose A, R, or M.")

    def publish_content(self, content: str, content_type: str):
        """Simulate publishing content."""
        print_section("📤 PUBLISHING CONTENT")
        print(f"Publishing {content_type} content...")
        # In a real scenario, this would publish to a blog, send an email, post to social media, etc.
        print_success(f"{content_type.upper()} content published successfully!")

        # Log the publication
        self.decision_log.append({
            "action": "published",
            "content_type": content_type,
            "content": content
        })

    def run_workflow(self, content_type: str, topic: str) -> bool:
        """
        Run the complete HITL workflow.

        Returns:
            bool: True if content was published, False otherwise
        """
        print_header("BASIC HITL CONTENT GENERATION WORKFLOW")

        print_section("📝 CONTENT GENERATION")
        print(f"Content Type: {content_type}")
        print(f"Topic: {topic}")
        print("\nGenerating content...")

        # Generate initial content
        additional_instructions = ""
        max_attempts = 3
        attempt = 1

        while attempt <= max_attempts:
            content = self.generator.generate_content(content_type, topic, additional_instructions)

            if not content:
                print_error("Failed to generate content. Aborting workflow.")
                return False

            print_success(f"Content generated (Attempt {attempt}/{max_attempts})")

            # Human checkpoint
            decision, feedback = self.get_human_decision(content, content_type)

            # Log the decision
            self.decision_log.append({
                "attempt": attempt,
                "decision": decision,
                "feedback": feedback,
                "content": content
            })

            if decision == 'approve':
                self.publish_content(content, content_type)
                return True

            elif decision == 'reject':
                print_section("❌ CONTENT REJECTED")
                print(f"Reason: {feedback}")
                print_error("Workflow terminated by human reviewer.")
                return False

            elif decision == 'modify':
                print_section("🔄 REGENERATING CONTENT")
                print(f"Feedback: {feedback}")
                additional_instructions = f"Previous feedback: {feedback}. Please incorporate this feedback."
                attempt += 1

                if attempt <= max_attempts:
                    print(f"\nRegenerating content (Attempt {attempt}/{max_attempts})...")
                else:
                    print_warning(f"Maximum attempts ({max_attempts}) reached.")
                    print_error("Workflow terminated.")
                    return False

        return False

    def show_audit_trail(self):
        """Display the audit trail of decisions."""
        print_section("📋 AUDIT TRAIL")

        if not self.decision_log:
            print("No decisions recorded.")
            return

        for i, entry in enumerate(self.decision_log, 1):
            print(f"\n{Colors.BOLD}Entry {i}:{Colors.ENDC}")
            for key, value in entry.items():
                if key == "content":
                    print(f"  {key}: [Content omitted for brevity]")
                else:
                    print(f"  {key}: {value}")


def main():
    """Main function to demonstrate the HITL workflow."""
    # Load environment variables
    env_path = Path(__file__).parent.parent.parent.parent / '.env'
    load_dotenv(env_path)

    api_key = os.getenv('OPENAI_API_KEY')
    if not api_key:
        print_error("OPENAI_API_KEY not found in environment variables.")
        print(f"Please ensure .env file exists at: {env_path}")
        sys.exit(1)

    # Initialize components
    generator = ContentGenerator(api_key)
    workflow = HITLWorkflow(generator)

    # Example scenarios
    scenarios = [
        ("blog", "The benefits of adopting a human-in-the-loop approach in AI systems"),
        ("email", "Announcing a new AI safety feature to customers"),
        ("social", "Celebrating our company's commitment to ethical AI"),
    ]

    print_header("WELCOME TO BASIC HITL DEMO")
    print("\nThis demo shows a simple human-in-the-loop workflow for content generation.")
    print("You'll be asked to approve, reject, or request modifications to generated content.\n")

    # Let user choose a scenario
    print_section("Available Scenarios")
    for i, (content_type, topic) in enumerate(scenarios, 1):
        print(f"{i}. {Colors.BOLD}{content_type.upper()}{Colors.ENDC}: {topic}")

    while True:
        try:
            choice = input(f"\n{Colors.BOLD}Select a scenario (1-{len(scenarios)}) or 'q' to quit:{Colors.ENDC} ").strip()

            if choice.lower() == 'q':
                print("\nGoodbye!")
                sys.exit(0)

            choice_idx = int(choice) - 1
            if 0 <= choice_idx < len(scenarios):
                content_type, topic = scenarios[choice_idx]
                break
            else:
                print_error(f"Please enter a number between 1 and {len(scenarios)}.")
        except ValueError:
            print_error("Invalid input. Please enter a number or 'q'.")

    # Run the workflow
    success = workflow.run_workflow(content_type, topic)

    # Show audit trail
    workflow.show_audit_trail()

    # Summary
    print_section("📊 WORKFLOW SUMMARY")
    if success:
        print_success("Content was successfully published!")
    else:
        print_error("Content was not published.")

    print(f"\nTotal decisions made: {len([e for e in workflow.decision_log if 'decision' in e])}")


if __name__ == "__main__":
    main()

File: foundational_design_patterns/10_hitl/typescript/src/hitl_basic.ts
/**
 * Basic Human-in-the-Loop (HITL) Implementation
 * — TypeScript port of src/hitl_basic.py.
 */

import OpenAI from "openai";
import { createInterface, type Interface } from "node:readline/promises";
import { stdin as input, stdout as output } from "node:process";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const DEFAULT_MODEL = "gpt-4o-mini";
const ENV_PATH = join(dirname(fileURLToPath(import.meta.url)), "../../../../.env");

function getDefaultModel(): string {
  return process.env.OPENAI_MODEL ?? DEFAULT_MODEL;
}

export class Colors {
  static HEADER = "\x1b[95m";
  static BLUE = "\x1b[94m";
  static CYAN = "\x1b[96m";
  static GREEN = "\x1b[92m";
  static YELLOW = "\x1b[93m";
  static RED = "\x1b[91m";
  static ENDC = "\x1b[0m";
  static BOLD = "\x1b[1m";
  static UNDERLINE = "\x1b[4m";
}

export function printHeader(text: string): void {
  console.log(`\n${Colors.HEADER}${Colors.BOLD}${"=".repeat(80)}${Colors.ENDC}`);
  console.log(`${Colors.HEADER}${Colors.BOLD}${text.padStart((80 + text.length) / 2).padEnd(80)}${Colors.ENDC}`);
  console.log(`${Colors.HEADER}${Colors.BOLD}${"=".repeat(80)}${Colors.ENDC}\n`);
}

export function printSection(title: string): void {
  console.log(`\n${Colors.CYAN}${Colors.BOLD}${title}${Colors.ENDC}`);
  console.log(`${Colors.CYAN}${"-".repeat(title.length)}${Colors.ENDC}`);
}

export function printContent(content: string): void {
  const lines = content.split("\n");
  const maxLength = Math.max(...lines.map((line) => line.length)) + 4;

  console.log(`\n${Colors.BLUE}+${"-".repeat(maxLength)}+${Colors.ENDC}`);
  for (const line of lines) {
    console.log(
      `${Colors.BLUE}|${Colors.ENDC} ${line.padEnd(maxLength - 2)} ${Colors.BLUE}|${Colors.ENDC}`,
    );
  }
  console.log(`${Colors.BLUE}+${"-".repeat(maxLength)}+${Colors.ENDC}\n`);
}

export function printSuccess(text: string): void {
  console.log(`${Colors.GREEN}${Colors.BOLD}✓ ${text}${Colors.ENDC}`);
}

export function printError(text: string): void {
  console.log(`${Colors.RED}${Colors.BOLD}✗ ${text}${Colors.ENDC}`);
}

export function printWarning(text: string): void {
  console.log(`${Colors.YELLOW}${Colors.BOLD}⚠ ${text}${Colors.ENDC}`);
}

export class ContentGenerator {
  client: OpenAI;

  constructor(apiKey: string) {
    this.client = new OpenAI({ apiKey });
  }

  async generateContent(
    contentType: string,
    topic: string,
    additionalInstructions = "",
  ): Promise<string | null> {
    const prompts: Record<string, string> = {
      blog: `Write a short blog post (200-300 words) about: ${topic}. ${additionalInstructions}`,
      email: `Write a professional email about: ${topic}. ${additionalInstructions}`,
      social: `Write a social media post (2-3 sentences) about: ${topic}. ${additionalInstructions}`,
      tweet: `Write a tweet (280 characters max) about: ${topic}. ${additionalInstructions}`,
    };

    const prompt = prompts[contentType] ?? prompts.blog ?? "";

    try {
      const response = await this.client.chat.completions.create({
        model: getDefaultModel(),
        messages: [
          {
            role: "system",
            content:
              "You are a professional content writer. Create engaging, clear, and concise content.",
          },
          { role: "user", content: prompt },
        ],
        temperature: 0.7,
        max_tokens: 500,
      });

      return response.choices[0]?.message?.content?.trim() ?? null;
    } catch (error: unknown) {
      printError(
        `Failed to generate content: ${
          error instanceof Error ? error.message : String(error)
        }`,
      );
      return null;
    }
  }
}

export type ReviewDecision = "approve" | "reject" | "modify";
export type AskFn = (prompt: string) => Promise<string>;

export interface DecisionEntry {
  attempt?: number;
  decision?: ReviewDecision;
  feedback?: string | null;
  content?: string;
  action?: string;
  content_type?: string;
}

export class HITLWorkflow {
  generator: ContentGenerator;
  decisionLog: DecisionEntry[];
  autoDecision: ReviewDecision | null;
  autoFeedback: string | null;
  ask: AskFn;

  constructor(
    generator: ContentGenerator,
    ask: AskFn,
    autoDecision?: ReviewDecision,
    autoFeedback?: string,
  ) {
    this.generator = generator;
    this.decisionLog = [];
    this.ask = ask;
    this.autoDecision = autoDecision ?? null;
    this.autoFeedback = autoFeedback ?? null;
  }

  async getHumanDecision(
    content: string,
    contentType: string,
  ): Promise<[ReviewDecision, string | null]> {
    printSection("🔍 HUMAN REVIEW CHECKPOINT");
    console.log(`\n${Colors.BOLD}Content Type:${Colors.ENDC} ${contentType.toUpperCase()}`);
    console.log(`${Colors.BOLD}Generated Content:${Colors.ENDC}`);
    printContent(content);

    console.log(`\n${Colors.YELLOW}${Colors.BOLD}Review Options:${Colors.ENDC}`);
    console.log(`  ${Colors.GREEN}[A]${Colors.ENDC} Approve - Publish the content as-is`);
    console.log(`  ${Colors.RED}[R]${Colors.ENDC} Reject - Discard the content`);
    console.log(`  ${Colors.YELLOW}[M]${Colors.ENDC} Modify - Request changes to the content`);

    if (this.autoDecision === "approve") {
      printWarning("Automatic review decision enabled: approve");
      return ["approve", this.autoFeedback];
    }
    if (this.autoDecision === "reject") {
      printWarning("Automatic review decision enabled: reject");
      return ["reject", this.autoFeedback ?? "Auto-rejected"];
    }
    if (this.autoDecision === "modify") {
      printWarning("Automatic review decision enabled: modify");
      return [
        "modify",
        this.autoFeedback ?? "Please tighten the draft and improve clarity.",
      ];
    }

    while (true) {
      const decision = (await this.ask(
        `\n${Colors.BOLD}Your decision [A/R/M]:${Colors.ENDC} `,
      ))
        .trim()
        .toUpperCase();

      if (decision === "A" || decision === "APPROVE") {
        return ["approve", null];
      }
      if (decision === "R" || decision === "REJECT") {
        const reason = await this.ask(
          `${Colors.BOLD}Reason for rejection (optional):${Colors.ENDC} `,
        );
        return ["reject", reason.trim() || "No reason provided"];
      }
      if (decision === "M" || decision === "MODIFY") {
        const feedback = (await this.ask(
          `${Colors.BOLD}What changes would you like?${Colors.ENDC} `,
        )).trim();
        if (feedback) {
          return ["modify", feedback];
        }
        printWarning("Please provide feedback for modifications.");
        continue;
      }
      printError("Invalid option. Please choose A, R, or M.");
    }
  }

  publishContent(content: string, contentType: string): void {
    printSection("📤 PUBLISHING CONTENT");
    console.log(`Publishing ${contentType} content...`);
    printSuccess(`${contentType.toUpperCase()} content published successfully!`);

    this.decisionLog.push({
      action: "published",
      content_type: contentType,
      content,
    });
  }

  async runWorkflow(contentType: string, topic: string): Promise<boolean> {
    printHeader("BASIC HITL CONTENT GENERATION WORKFLOW");

    printSection("📝 CONTENT GENERATION");
    console.log(`Content Type: ${contentType}`);
    console.log(`Topic: ${topic}`);
    console.log("\nGenerating content...");

    let additionalInstructions = "";
    const maxAttempts = 3;
    let attempt = 1;

    while (attempt <= maxAttempts) {
      const content = await this.generator.generateContent(
        contentType,
        topic,
        additionalInstructions,
      );

      if (!content) {
        printError("Failed to generate content. Aborting workflow.");
        return false;
      }

      printSuccess(`Content generated (Attempt ${attempt}/${maxAttempts})`);
      const [decision, feedback] = await this.getHumanDecision(content, contentType);

      this.decisionLog.push({
        attempt,
        decision,
        feedback,
        content,
      });

      if (decision === "approve") {
        this.publishContent(content, contentType);
        return true;
      }

      if (decision === "reject") {
        printSection("❌ CONTENT REJECTED");
        console.log(`Reason: ${feedback}`);
        printError("Workflow terminated by human reviewer.");
        return false;
      }

      printSection("🔄 REGENERATING CONTENT");
      console.log(`Feedback: ${feedback}`);
      additionalInstructions =
        `Previous feedback: ${feedback}. Please incorporate this feedback.`;
      attempt += 1;

      if (attempt <= maxAttempts) {
        console.log(`\nRegenerating content (Attempt ${attempt}/${maxAttempts})...`);
      } else {
        printWarning(`Maximum attempts (${maxAttempts}) reached.`);
        printError("Workflow terminated.");
        return false;
      }
    }

    return false;
  }

  showAuditTrail(): void {
    printSection("📋 AUDIT TRAIL");

    if (!this.decisionLog.length) {
      console.log("No decisions recorded.");
      return;
    }

    this.decisionLog.forEach((entry, index) => {
      console.log(`\n${Colors.BOLD}Entry ${index + 1}:${Colors.ENDC}`);
      Object.entries(entry).forEach(([key, value]) => {
        if (key === "content") {
          console.log(`  ${key}: [Content omitted for brevity]`);
        } else {
          console.log(`  ${key}: ${value}`);
        }
      });
    });
  }
}

export interface BasicArgs {
  scenario?: number;
  autoDecision?: ReviewDecision;
  autoFeedback?: string;
}

export function parseBasicArgs(argv: string[]): BasicArgs {
  const args: BasicArgs = {};

  for (let index = 0; index < argv.length; index += 1) {
    const arg = argv[index];
    if (arg === "--scenario") {
      args.scenario = Number(argv[index + 1]);
      index += 1;
      continue;
    }
    if (arg === "--auto-decision") {
      const value = argv[index + 1] as ReviewDecision | undefined;
      if (value) {
        args.autoDecision = value;
      }
      index += 1;
      continue;
    }
    if (arg === "--auto-feedback") {
      args.autoFeedback = argv[index + 1];
      index += 1;
    }
  }

  return args;
}

async function askWithInterface(rl: Interface, prompt: string): Promise<string> {
  return await rl.question(prompt);
}

export async function main(argv = process.argv.slice(2)): Promise<number> {
  const args = parseBasicArgs(argv);
  const apiKey = process.env.OPENAI_API_KEY;
  if (!apiKey) {
    printError("OPENAI_API_KEY not found in environment variables.");
    console.log(`Please ensure .env file exists at: ${ENV_PATH}`);
    return 1;
  }

  const rl = createInterface({ input, output });
  try {
    const generator = new ContentGenerator(apiKey);
    const workflow = new HITLWorkflow(
      generator,
      (prompt) => askWithInterface(rl, prompt),
      args.autoDecision,
      args.autoFeedback,
    );

    const scenarios: Array<[string, string]> = [
      [
        "blog",
        "The benefits of adopting a human-in-the-loop approach in AI systems",
      ],
      ["email", "Announcing a new AI safety feature to customers"],
      ["social", "Celebrating our company's commitment to ethical AI"],
    ];

    printHeader("WELCOME TO BASIC HITL DEMO");
    console.log(
      "\nThis demo shows a simple human-in-the-loop workflow for content generation.",
    );
    console.log(
      "You'll be asked to approve, reject, or request modifications to generated content.\n",
    );

    printSection("Available Scenarios");
    scenarios.forEach(([contentType, topic], index) => {
      console.log(
        `${index + 1}. ${Colors.BOLD}${contentType.toUpperCase()}${Colors.ENDC}: ${topic}`,
      );
    });

    let contentType: string;
    let topic: string;

    if (args.scenario == null) {
      while (true) {
        const choice = (await askWithInterface(
          rl,
          `\n${Colors.BOLD}Select a scenario (1-${scenarios.length}) or 'q' to quit:${Colors.ENDC} `,
        )).trim();

        if (choice.toLowerCase() === "q") {
          console.log("\nGoodbye!");
          return 0;
        }

        const choiceIndex = Number(choice) - 1;
        if (Number.isInteger(choiceIndex) && choiceIndex >= 0 && choiceIndex < scenarios.length) {
          [contentType, topic] = scenarios[choiceIndex]!;
          break;
        }

        printError(`Please enter a number between 1 and ${scenarios.length}.`);
      }
    } else {
      [contentType, topic] = scenarios[args.scenario - 1]!;
      printWarning(
        `Scenario override enabled. Auto-selecting scenario ${args.scenario}: ${contentType.toUpperCase()} - ${topic}`,
      );
    }

    const success = await workflow.runWorkflow(contentType, topic);
    workflow.showAuditTrail();

    printSection("📊 WORKFLOW SUMMARY");
    if (success) {
      printSuccess("Content was successfully published!");
    } else {
      printError("Content was not published.");
      if (args.scenario != null || args.autoDecision != null) {
        return 1;
      }
    }

    const decisionCount = workflow.decisionLog.filter((entry) => entry.decision).length;
    console.log(`\nTotal decisions made: ${decisionCount}`);
    return success ? 0 : 1;
  } finally {
    rl.close();
  }
}

if (import.meta.main) {
  process.exitCode = await main();
}
